File: vane-permissions/src/main/java/org/oddlama/vane/permissions/argumentTypes/PermissionGroupArgumentType.java
package org.oddlama.vane.permissions.argumentTypes;

import com.mojang.brigadier.arguments.ArgumentType;
import com.mojang.brigadier.arguments.StringArgumentType;
import com.mojang.brigadier.context.CommandContext;
import com.mojang.brigadier.exceptions.CommandSyntaxException;
import com.mojang.brigadier.suggestion.Suggestions;
import com.mojang.brigadier.suggestion.SuggestionsBuilder;
import io.papermc.paper.command.brigadier.argument.CustomArgumentType;
import java.util.Map;
import java.util.Set;
import java.util.concurrent.CompletableFuture;
import java.util.stream.Stream;
import org.jetbrains.annotations.NotNull;

public class PermissionGroupArgumentType implements CustomArgumentType.Converted<String, String> {

    Map<String, Set<String>> permissionGroups;

    public static @NotNull PermissionGroupArgumentType permissionGroup(Map<String, Set<String>> permissionGroups) {
        return new PermissionGroupArgumentType(permissionGroups);
    }

    private PermissionGroupArgumentType(Map<String, Set<String>> permissionGroups) {
        this.permissionGroups = permissionGroups;
    }

    @Override
    public @NotNull String convert(@NotNull String nativeType) throws CommandSyntaxException {
        return nativeType;
    }

    @Override
    public @NotNull ArgumentType<String> getNativeType() {
        return StringArgumentType.string();
    }

    @Override
    public <S> @NotNull CompletableFuture<Suggestions> listSuggestions(
        @NotNull CommandContext<S> context,
        @NotNull SuggestionsBuilder builder
    ) {
        Stream<String> stream = permissionGroups.keySet().stream();
        if (!builder.getRemaining().isBlank()) {
            stream = stream.filter(group -> group.contains(builder.getRemaining()));
        }
        stream.forEach(builder::suggest);
        return builder.buildFuture();
    }
}

File: vane-permissions/src/main/kotlin/org/oddlama/vane/permissions/argumentTypes/PermissionGroupArgumentType.kt
package org.oddlama.vane.permissions.argumentTypes

import com.mojang.brigadier.arguments.ArgumentType
import com.mojang.brigadier.arguments.StringArgumentType
import com.mojang.brigadier.context.CommandContext
import com.mojang.brigadier.exceptions.CommandSyntaxException
import com.mojang.brigadier.suggestion.Suggestions
import com.mojang.brigadier.suggestion.SuggestionsBuilder
import io.papermc.paper.command.brigadier.argument.CustomArgumentType
import java.util.concurrent.CompletableFuture

/**
 * Brigadier argument type that validates and suggests configured permission group names.
 *
 * @property permissionGroups available groups used for command suggestions.
 */
class PermissionGroupArgumentType private constructor(private val permissionGroups: Map<String, Set<String>>) :
    CustomArgumentType.Converted<String, String> {
    /** Returns the parsed permission group name unchanged. */
    @Throws(CommandSyntaxException::class)
    override fun convert(nativeType: String): String {
        return nativeType
    }

    /** Uses Brigadier string parsing for the underlying native type. */
    override fun getNativeType(): ArgumentType<String> = StringArgumentType.string()

    /** Suggests known group names that contain the currently typed input. */
    override fun <S : Any> listSuggestions(
        context: CommandContext<S>,
        builder: SuggestionsBuilder
    ): CompletableFuture<Suggestions> {
        val remaining = builder.remaining
        permissionGroups.keys
            .asSequence()
            .filter { remaining.isBlank() || it.contains(remaining) }
            .forEach(builder::suggest)
        return builder.buildFuture()
    }

    /** Static factories for permission group argument creation. */
    companion object {
        /** Factory for creating a permission group argument backed by current group configuration. */
        fun permissionGroup(permissionGroups: MutableMap<String, MutableSet<String>>): PermissionGroupArgumentType =
            PermissionGroupArgumentType(permissionGroups)
    }
}
